kubelet: enable crio

rphillips opened the pull request.

abhinavdahiya commented on templates/_base/master/files/container-registries.yaml:
```diff
@@ -0,0 +1,30 @@
+filesystem: "root"
```
@rphillips 
* why is this file required?
* who uses this ?
* how will we configure it if we somebody needs to configure it for their internal registry ??

/cc @crawford

rphillips replied:
The file is needed since crio gets installed with a default config which only includes the `registry.access.redhat.com` registry. docker.io needs to be added to the registry list to pull in defaulted docker images.

/cc @sjenning

rphillips replied:
updated

abhinavdahiya commented:
@rphillips also can you run `go test ./pkg/controller/template/... -u` to update the golden test files.

rphillips commented:
@abhinavdahiya regenerated

rphillips commented:
/hold

We may want to merge the cross-repo crio PR's all at once.

crawford commented on templates/_base/master/files/container-registries.yaml:
```diff
@@ -0,0 +1,30 @@
+
+    # If you need to access insecure registries, add the registry's fully-qualified name.
+    # An insecure registry is one that does not have a valid SSL certificate or only does HTTP.
+    [registries.insecure]
```
As mentioned in https://github.com/openshift/installer/pull/234#discussion_r216755364, can you remove these empty sections?

rphillips replied:
updated

wking commented on pkg/controller/template/test_data/templates/aws/master/units/kubelet.service:
```diff
@@ -12,6 +12,9 @@ contents: |
         --rotate-certificates \
+        --container-runtime=remote \
+        --container-runtime-endpoint=unix:///var/run/crio/crio.sock \
+        --runtime-request-timeout=10m \
```
Why did you need to set this?  Did you hit an error without it, or is it just increasing robustness?  What's the default value?  How did you decide on 10m?

sjenning replied:
default is 2m.  probably doesn't need to be changed.  i think the cri-o docs recommend this, but i don't think it is needed.

wking replied:
> default is 2m. probably doesn't need to be changed

I'm in favor of dropping it then.  The fewer local opinions we need to maintain, the better :).

rphillips replied:
Yeah, I took that from this recommendation: https://github.com/kubernetes-sigs/cri-o/blob/master/kubernetes.md#preparing-kubelet

abhinavdahiya replied:
Let's keep the time out then, as it might be required when we pull hyberkube image on slower network.

wking commented on templates/_base/worker/files/container-registries.yaml:
```diff
@@ -0,0 +1,30 @@
+filesystem: "root"
```
It DRY things up a bit to have a `templates/_base/all/files/container-registries.yaml` or some such which would be added to both masters and workers.  But perhaps that is more than we want to bite off in this particular PR.

sjenning commented on pkg/controller/template/test_data/templates/aws/worker/units/kubelet.service:
```diff
@@ -12,6 +12,9 @@ contents: |
         --rotate-certificates \
+        --container-runtime=remote \
+        --container-runtime-endpoint=unix:///var/run/crio/crio.sock \
+        --runtime-request-timeout=10m \
```
same here

rphillips commented:
Updated the PR and removed the --runtime-request-timeout.

The default --runtime-request-timeout enforces a 2 minute timeout on image pulls... We may want to bump it up for slower networks.

rphillips commented:
/hold cancel

abhinavdahiya commented on templates/_base/worker/units/kubelet.yaml:
```diff
@@ -14,6 +14,8 @@ contents: |
         --rotate-certificates \
+        --container-runtime=remote \
+        --container-runtime-endpoint=unix:///var/run/crio/crio.sock \
         --cni-conf-dir=/etc/kubernetes/cni/net.d \
```
`--cni-conf-dir`, `--cni-bin-dir`, `--network-pluging` are no longer required as cri-o handles everything.

rphillips replied:
Thanks. Removed those options.

abhinavdahiya commented:
/hold

I was testing these changes yesterday, networking is not ready to be used with crio in installer yet. https://github.com/openshift/installer/pull/235#issuecomment-420854934

abhinavdahiya commented:
> The default --runtime-request-timeout enforces a 2 minute timeout on image pulls... We may want to bump it up for slower networks.

As it is recommended [here](The default --runtime-request-timeout enforces a 2 minute timeout on image pulls... We may want to bump it up for slower networks.) Let's set it to 10m

rphillips commented:
@abhinavdahiya I added support for the default 10m timeout, and an environment variable and file to be able to overwrite the default

sjenning commented:
I think all the feedback has been addressed.  I've tested this an it works:
```
$ oc get nodes -o wide
NAME            STATUS    ROLES       AGE       VERSION           INTERNAL-IP      EXTERNAL-IP   OS-IMAGE                  KERNEL-VERSION               CONTAINER-RUNTIME
dev-bootstrap   Ready     bootstrap   5m        v1.11.0+d4cacc0   192.168.124.10   <none>        Red Hat CoreOS 4.0.5470   3.10.0-862.14.1.el7.x86_64   cri-o://1.11.2
dev-master-0    Ready     master      4m        v1.11.0+d4cacc0   192.168.124.11   <none>        Red Hat CoreOS 4.0.5470   3.10.0-862.14.1.el7.x86_64   cri-o://1.11.2
dev-worker-0    Ready     worker      4m        v1.11.0+d4cacc0   192.168.124.50   <none>        Red Hat CoreOS 4.0.5470   3.10.0-862.14.1.el7.x86_64   cri-o://1.11.2
dev-worker-1    Ready     worker      4m        v1.11.0+d4cacc0   192.168.124.51   <none>        Red Hat CoreOS 4.0.5470   3.10.0-862.14.1.el7.x86_64   cri-o://1.11.2
```
can this get approved?

wking commented:
This looks good to me.  @abhinavdahiya, do you want to mark your requested-change resolved?

abhinavdahiya commented:
@wking https://github.com/openshift/machine-config-operator/pull/52#issuecomment-421019923

abhinavdahiya commented:
@rphillips @sjenning 
Can you add these files to the PR:
```console
$ cat templates/_base/master/files/etc-sysconfig-crio-network.yaml
filesystem: "root"
mode: 0644
path: "/etc/sysconfig/crio-network"
contents:
  inline: |
    CRIO_NETWORK_OPTIONS=--cni-config-dir=/etc/kubernetes/cni/net.d --cni-plugin-dir=/etc/kubernetes/cni/bin

$ cat templates/_base/worker/files/etc-sysconfig-crio-network.yaml
filesystem: "root"
mode: 0644
path: "/etc/sysconfig/crio-network"
contents:
  inline: |
    CRIO_NETWORK_OPTIONS=--cni-config-dir=/etc/kubernetes/cni/net.d --cni-plugin-dir=/etc/kubernetes/cni/bin
``

Also rebase to 2 commits, one with changes and one with generated test_data `go test ./pkg/controller/templates/.. -u`
